mpc-core/src/protocols/plain.rs

```diff
@@ -637,51 +637,78 @@ impl<F: PrimeField> NoirWitnessExtensionProtocol<F> for PlainDriver<F> {
         Ok(-c / q_l)
     }
 
-    fn read_lut_by_acvm_type(&mut self, index: &Self::AcvmType, lut: &Self::LUT) -> F {
+    fn read_lut_by_acvm_type(
+        &mut self,
+        index: &Self::AcvmType,
+        lut: &Self::SecretSharedMap,
+    ) -> eyre::Result<F> {
         self.get_from_lut(index, lut)
     }
 
     fn write_lut_by_acvm_type(
         &mut self,
         index: Self::AcvmType,
         value: Self::AcvmType,
-        lut: &mut Self::LUT,
-    ) {
-        self.write_to_lut(index, value, lut);
+        map: &mut Self::SecretSharedMap,
+    ) -> eyre::Result<()> {
+        self.write_to_lut(index, value, map)
     }
 
-    fn init_lut_by_acvm_type(&mut self, values: Vec<Self::AcvmType>) -> Self::LUT {
-        self.init_lut(values)
+    fn init_lut_by_acvm_type(&mut self, values: Vec<Self::AcvmType>) -> Self::SecretSharedMap {
+        self.init_map(values.into_iter().enumerate().map(|(idx, value)| {
+            let promoted_idx = self.promote_to_trivial_share(F::from(
+                u64::try_from(idx).expect("usize fits into u64"),
+            ));
+            (promoted_idx, value)
+        }))
     }
 }
 
 impl<F: PrimeField> LookupTableProvider<F> for PlainDriver<F> {
-    type LUT = HashMap<F, F>;
-
-    fn init_lut(&mut self, values: Vec<F>) -> Self::LUT {
-        let mut lut = HashMap::with_capacity(values.len());
-        for (idx, value) in values.into_iter().enumerate() {
-            lut.insert(
-                F::from(u64::try_from(idx).expect("usize fits into u64")),
-                value,
-            );
-        }
-        lut
+    // we could check if a Vec<F> impl may be faster. Depends on the size of the LUT..
+    type SecretSharedSet = HashSet<F>;
+
+    type SecretSharedMap = HashMap<F, F>;
+
```

Comment:
I think we would need `Vec`s here for correctness anyway, since just because the share is equal does not mean the underling value is equal.

Reply:
Also for the plain value driver? The "shares" are F. For the rep3 one we have Vecs.